test/bio_enc_test.c

```diff
@@ -64,53 +88,61 @@ int main()
         int delta;
 
         b = BIO_new(BIO_f_cipher());
-        if (!BIO_set_cipher(b, EVP_aes_128_cbc(), key, NULL, 0))
-             return -1;
-        BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
+        if (!TEST_true(BIO_set_cipher(b, cipher, key, iv, ENCRYPT))) {
+            TEST_info("Small chunk encrypt failed @ operation %d", i);
+            return 0;
+        }
+        BIO_push(b, BIO_new_mem_buf(inp, DATA_SIZE));
         memset(out, 0, sizeof(out));
         for (len = 0; (delta = BIO_read(b, out + len, i)); ) {
             len += delta;
         }
         BIO_free_all(b);
 
-        if (len != lref || memcmp(out, ref, len)) {
-            fprintf(stderr, "CBC output mismatch@%d\n", i);
-            return 3;
+        if (!TEST_mem_eq(out, len, ref, lref)) {
+            TEST_info("Small chunk encrypt compare failed @ operation %d", i);
+            return 0;
         }
     }
 
-    /*
-     * Exercise CTR cipher
-     */
+    /* Decrypt tests */
 
     /* reference output for single-chunk operation */
     b = BIO_new(BIO_f_cipher());
-    if (!BIO_set_cipher(b, EVP_aes_128_ctr(), key, NULL, 0))
-         return -1;
-    BIO_push(b, BIO_new_mem_buf(inp, sizeof(inp)));
-    lref = BIO_read(b, ref, sizeof(ref));
+    if (!TEST_true(BIO_set_cipher(b, cipher, key, iv, DECRYPT)))
+        return 0;
+    /* Use original reference output as input */
+    BIO_push(b, BIO_new_mem_buf(ref, lref));
+    (void)BIO_flush(b);
```

Comment:
Is there a reason for this `BIO_flush()` it is by now in OpenSSL 3.x definitely not correct, but was there a plausible reason for this back when introduced?

Reply:
No idea, it's been some time since I wrote this. I agree it doesn't need to be there though.